### PR TITLE
SAK-42608 Samigo display user notifications for available assessments

### PR DESCRIPTION
**Pull request by jkjanetschek**
This implementation does consider delays and exceptions for users/groups

### CODE REVIEW

File: samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/author/SavePublishedSettingsListener.java [hunk 1]
```diff
@@ -239,6 +253,61 @@ public void processAction(ActionEvent ae) throws AbortProcessingException
...
+				while (newtimes.hasNext()){
+					ExtendedTime newExTime = (ExtendedTime) newtimes.next();
+					if(newExTime.getStartDate().toInstant().isAfter(Instant.now())){
+						eventTrackingService.delay(eventTrackingService.newEvent(SamigoConstants.EVENT_ASSESSMENT_AVAILABLE, "siteId=" + AgentFacade.getCurrentSiteId() + ", assessmentId=" + assessmentSettings.getAssessmentId() + ", publishedAssessmentId=" + assessment.getPublishedAssessmentId(), true), newExTime.getStartDate().toInstant());
+					}
+				}
+			}else {
+				ListIterator<ExtendedTime> oldtimes = oldExtendedTimes.listIterator();
+				ListIterator<ExtendedTime> newtimes = assessmentSettings.getExtendedTimes().listIterator();
+				while(oldtimes.hasNext()){
```
Comment (adrianfish) [inline samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/author/SavePublishedSettingsListener.java:292]:
```suggestion
				while (oldtimes.hasNext()) {
```

File: samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/author/SavePublishedSettingsListener.java [hunk 2]
```diff
@@ -239,6 +253,61 @@ public void processAction(ActionEvent ae) throws AbortProcessingException
...
+					ExtendedTime oldExTime = (ExtendedTime) oldtimes.next();
+					while (newtimes.hasNext()){
```
Comment (adrianfish) [inline samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/author/SavePublishedSettingsListener.java:294]:
```suggestion
					while (newtimes.hasNext()) {
```

File: samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/author/SavePublishedSettingsListener.java [hunk 3]
```diff
@@ -239,6 +253,61 @@ public void processAction(ActionEvent ae) throws AbortProcessingException
...
+						ExtendedTime newExTime = (ExtendedTime) newtimes.next();
+						if(!newExTime.equals(oldExTime)){
```
Comment (adrianfish) [inline samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/author/SavePublishedSettingsListener.java:296]:
```suggestion
						if (!newExTime.equals(oldExTime)) {
```

File: samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/author/SavePublishedSettingsListener.java [hunk 4]
```diff
@@ -239,6 +253,61 @@ public void processAction(ActionEvent ae) throws AbortProcessingException
...
+							if(!flag){
+								eventTrackingService.post(eventTrackingService.newEvent(SamigoConstants.EVENT_ASSESSMENT_UPDATE_AVAILABLE, "siteId=" + AgentFacade.getCurrentSiteId() + ", assessmentId=" + assessment.getAssessmentId() + ", publishedAssessmentId=" + assessment.getPublishedAssessmentId(), true));
+								flag = true;
+							}
+							if(newExTime.getStartDate().toInstant().isAfter(Instant.now())){
```
Comment (adrianfish) [inline samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/author/SavePublishedSettingsListener.java:301]:
```suggestion
							if (newExTime.getStartDate().toInstant().isAfter(Instant.now())) {
```

File: samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/author/RepublishAssessmentListener.java [hunk 5]
```diff
@@ -173,6 +183,39 @@ public void processAction(ActionEvent ae) throws AbortProcessingException {
...
+		} else {
+			eventTrackingService.delay(eventTrackingService.newEvent(SamigoConstants.EVENT_ASSESSMENT_AVAILABLE, "siteId=" + AgentFacade.getCurrentSiteId() + ", assessmentId=" + assessment.getAssessmentId() + ", publishedAssessmentId=" + assessment.getPublishedAssessmentId(), true), instant);
+			if (publishedAssessmentSettings.getExtendedTimesSize() != 0) {
+				ListIterator<ExtendedTime> it = extendedTimes.listIterator();
+				while (it.hasNext()) {
+					ExtendedTime exTime = (ExtendedTime) it.next();
+					Instant startInstant = exTime.getStartDate().toInstant();
+					if (startInstant.isBefore(Instant.now())) {
+						eventTrackingService.post(eventTrackingService.newEvent(SamigoConstants.EVENT_ASSESSMENT_UPDATE_AVAILABLE, "siteId=" + AgentFacade.getCurrentSiteId() + ", assessmentId=" + assessment.getAssessmentId() + ", publishedAssessmentId=" + assessment.getPublishedAssessmentId(), true));
+					} else if(startInstant.isAfter(Instant.now()) && !instant.equals(startInstant)) {
```
Comment (adrianfish) [inline samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/author/RepublishAssessmentListener.java:212]:
```suggestion
					} else if (startInstant.isAfter(Instant.now()) && !instant.equals(startInstant)) {
```

**Review by adrianfish — APPROVED** [review]
Thanks, that'll do. Building those references, drying out that part of the code, that can wait. Thanks.

**Comment by jkjanetschek** [comment]
Thanks for your input and review! 
Some repetition could easily be avoided, because many events in a code region use exactly the  same reference. I can make another commit to increase the number of reference reuse, where it makes sense. 

I am not sure how to approach a service method for reference building. For me it would make sense to place it in class AssessmentService. In a short search i found 6 possible references for events. Are there any missing?
Would it make sense to make just a service method with a switch-case  block?

EVENT_ASSESSMENT_AVAILABLE
EVENT_ASSESSMENT_DELETE
EVENT_ASSESSMENT_UPDATE_AVAILABLE 
EVENT_ASSESSMENT_PUBLISH
EVENT_PUBLISHED_ASSESSMENT_RETRACTED
      siteId=<siteId>, assessmentId=<assessmentId>, publishedAssessmentId=<publishedAssessmentId>
EVENT_PUBLISHED_ASSESSMENT_REPUBLISH 
EVENT_PUBLISHED_ASSESSMENT_SETTING_EDIT
EVENT_PUBLISHED_ASSESSMENT_REMOVE
       siteId=<siteId>, publishedAssessmentId=<publishedAssessmentId>
EVENT_ASSESSMENT_REMOVE
      assessmentId=<assessmentId>
EVENT_PUBLISHED_ASSESSMENT_SAVEITEM
      /sam/<siteId>/publish, publishedItemId=<publishedItemId>
EVENT_ASSESSMENT_UNINDEXITEM
      /sam/<siteId>/unindexed, itemId=<ItemId>
EVENT_PUBLISHED_ASSESSMENT_UNINDEXITEM
       sam/<siteId>/unindexed, publishedItemId=<publishedItemId>

**Comment by adrianfish** [comment]
Don't worry about it for now. It's a refactor we can do later. Let's get your change in so it can be tested out. Thanks for the work - the notifications stuff has quite a few moving parts.

**Pull request merged**